tez-dag/src/main/java/org/apache/tez/dag/app/dag/impl/TaskImpl.java

```diff
@@ -1009,14 +1014,18 @@ public TaskStateInternal transition(TaskImpl task, TaskEvent event) {
     public void transition(TaskImpl task, TaskEvent event) {
       LOG.info("Scheduling a redundant attempt for task " + task.taskId);
       task.counters.findCounter(TaskCounter.NUM_SPECULATIONS).increment(1);
-      TezTaskAttemptID earliestUnfinishedAttempt = null;
+      TaskAttempt earliestUnfinishedAttempt = null;
       for (TaskAttempt ta : task.attempts.values()) {
         // find the oldest running attempt
         if (!ta.isFinished()) {
-          earliestUnfinishedAttempt = ta.getID();
+          earliestUnfinishedAttempt = ta;
         }
       }
-      task.addAndScheduleAttempt(earliestUnfinishedAttempt);
+      NodeId nodeId = earliestUnfinishedAttempt.getNodeId();
+      task.unhealthyNodesHistory.add(nodeId);
```

Comment:
a piece of thought for discussions: in the loop above (line 1018-1021), should we put the node id of ALL running attempts here into unhealthy nodes (instead of just the earliest attempt)?  since by the time we got here, a redundant attempt was deemed necessary, that must be suggesting all the unfinished(running) attempts for this task are potentially problematic/long-tailed.

Reply:
for now there would be only one speculative attempt for one task, so maybe putting the node ids of all running attempts is unnecessary?

Reply:
Indeed, the default speculator enforces that only one speculative attempt is allowed today, but i would still think including all running nodes here would be slightly better:

1. with only one speculation allowed, there can be one running attempts above when we got here, so just adding its node to unhealthyNodeHistory is equivalent to adding the earliestUnfinishedAttempt's node.
2.  the fact that there can be only one speculative today, should be orthogonal to our desire to avoid using the same (possibly problematic) nodes for speculative attempt, therefore syntax wise, I think adding all running nodes (at the time when we are adding additional speculative attempt) to unhealthy history would make more sense, and it would be extensible too in the future, if we decide that having more than one speculative should be allowed.

Reply:
Thanks for your review, sounds very thoughtful, I'd add all running nodes to unhealthy history.